streams/src/main/java/org/apache/kafka/streams/processor/internals/assignment/SubscriptionInfo.java

```diff
@@ -200,10 +201,19 @@ public UUID processId() {
     public Map<TaskId, Long> taskOffsetSums() {
         if (taskOffsetSumsCache == null) {
             taskOffsetSumsCache = new HashMap<>();
-            for (final TaskOffsetSum topicGroup : data.taskOffsetSums()) {
-                for (final PartitionToOffsetSum partitionOffsetSum : topicGroup.partitionToOffsetSum()) {
-                    taskOffsetSumsCache.put(new TaskId(topicGroup.topicGroupId(), partitionOffsetSum.partition()),
-                                            partitionOffsetSum.offsetSum());
+            if (data.version() >= MIN_VERSION_OFFSET_SUM_SUBSCRIPTION) {
+                for (final TaskOffsetSum topicGroup : data.taskOffsetSums()) {
+                    for (final PartitionToOffsetSum partitionOffsetSum : topicGroup.partitionToOffsetSum()) {
+                        taskOffsetSumsCache.put(new TaskId(topicGroup.topicGroupId(), partitionOffsetSum.partition()),
+                            partitionOffsetSum.offsetSum());
+                    }
+                }
+            } else {
+                for (final TaskId task : prevTasks()) {
+                    taskOffsetSumsCache.put(task, Task.LATEST_OFFSET);
+                }
+                for (final TaskId task : standbyTasks()) {
+                    taskOffsetSumsCache.put(task, 0L);
```

Comment:
This now looks a little suspicious... In the absence of a value, should we assume standbys are caught up (`0L`), or that they are not (`MAX_VALUE`)?

Reply:
Hm, I guess if we do that then the assignment algorithm should hopefully reduce to the original one during an upgrade while we're pinned to the lower subscription version.

The assignment will be a little screwy on the first VP rebalance while there are mixed subscription versions, but maybe that's not worth worrying about. Probably not worth adding another sentinel value either..

Reply:
Actually I reverse, think we should add a (negative) `OFFSET_SUM_UNKNOWN` sentinel for this case. Either way we have to do some special handling, this way we can still enforce the `endOffsetSum >= taskOffsetSum` invariant and also don't share this pseudo-sentinel with the overflow case